src/fparser/two/Fortran2003.py

```diff
@@ -4909,39 +4909,193 @@ class Defined_Binary_Op(Base):  # pylint: disable=invalid-name
+
+    default-char-expr is expr
+
+    C707 default-char-expr shall be of type default character.
+
+    '''
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for a default character expression.
+
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
+
+        '''
+        excluded = (
+            Signed_Int_Literal_Constant, Int_Literal_Constant,
+            Binary_Constant, Octal_Constant, Hex_Constant,
+            Signed_Real_Literal_Constant, Real_Literal_Constant,
+            Complex_Literal_Constant, Logical_Literal_Constant)
+        # Attempt to match as a general expression.
+        result = Expr(string)
+        # C707: the match should fail if the result is not a character
+        # expression. This is difficult to check in general so for the
+        # time being check that, in the case where a literal constant
+        # is returned, this is not of the wrong type.
+        if isinstance(result, excluded):
+            return None
+        return result
+
+
+class Int_Expr(Base):  # pylint: disable=invalid-name
+    '''
+    Fortran 2003 rule R727
+
+    int-expr is expr
+
+    C708 int-expr shall be of type integer.
+
+    '''
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for an integer expression.
+
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
+
+        '''
+        excluded = (
+            Binary_Constant, Octal_Constant, Hex_Constant,
+            Signed_Real_Literal_Constant, Real_Literal_Constant,
+            Complex_Literal_Constant, Char_Literal_Constant,
+            Logical_Literal_Constant)
+        # Attempt to match as a general expression.
+        result = Expr(string)
+        # C708: the match should fail if the result is not an integer
+        # expression. This is difficult to check in general so for the
+        # time being check that, in the case where a literal constant
+        # is returned, this is not of the wrong type.
+        if isinstance(result, excluded):
+            return None
+        return result
+
+
+class Numeric_Expr(Base):  # pylint: disable=invalid-name
+    '''
+    Fortran 2003 rule R728
+
+    numeric-expr is expr
+
+    C709 numeric-expr shall be of type integer, real or complex.
+
+    '''
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for a numeric expression.
+
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
```

Comment:
Please complete and throughout.

Reply:
oops. Sorry, I forgot to add this then replicated with cut-n-paste.